packages/cojson-transport-ws/src/index.ts

```diff
@@ -69,21 +67,21 @@ export function createWebSocketPeer(options: {
             return true;
         };
 
-        yield* Queue.take(outgoing).pipe(
-            Effect.andThen((message) => ws.send(JSON.stringify(message))),
-            Effect.forever,
-            Effect.whenEffect(
-                fromEventListener(ws, "open").pipe(Effect.as(true)),
-            ),
-            Effect.forkDaemon, // :/
-        );
+        yield* Effect.gen(function* () {
+            yield* once(ws, "open");
+            yield* Queue.take(outgoing).pipe(
+                Effect.andThen((message) => ws.send(JSON.stringify(message))),
+                Effect.forever,
+                Effect.forkDaemon,
+            );
+        });
```

Comment:
wdyt about having a managed thingy?

```suggestion
        yield* Effect.gen(function* () {
            yield* once(ws, "open");
            yield* Queue.take(outgoing).pipe(
                Effect.andThen((message) => ws.send(JSON.stringify(message))),
                Effect.forever,
                Effect.forkScoped,
            );
            yield* once(ws, "close");
        }).pipe(Effect.scoped, Effect.forever, Effect.forkDaemon);
```

it might be ugly but what I'm trying to encapsulate is: dont send stuff to the websocket if it's not open, and wait for "open" once it's closed. Maybe it just means that we need to merge the outgoing into the messages stream?

Reply:
basically, managing it like that would mean that for `n` sockets created and `m<=n` being used, we're raising `n-m` zombies and `m` daemons...
Once the the socket would manage itself instead of needing consumers to manage it, this won't be an issue..
Basically, I'd recommend using `@effect/platform/Socket` and implementing [a client for it](https://github.com/Effect-TS/effect/blob/3429c8fb61dcef45df1d954c18b94851b6a9226f/packages/experimental/src/DevTools/Client.ts#L45). It's already generic, it's builtin, its duplex, and Tim would be maintaining it instead of us so we could focus on more important things. But until then... We don't need to expect more than a single `open` event 😄